repoaudit/repoaudit/apt.py

```diff
@@ -17,92 +19,252 @@
 CHUNK_SIZE = 2 * 1024 * 1024
 
 
-def _find_dists(url: str) -> Set[str]:
+def _find_dists(url: str, verify: Optional[str] = None) -> Set[str]:
+    """Find apt distributions."""
     try:
-        resp = get_url(urljoin(url, "dists"))
-    except HTTPError as e:
-        if e.response.status_code == 404:
-            raise click.ClickException(f"Could not determine dists from {url}."
-                                       "Please manually supply them with --dist.")
+        resp = get_url(urljoin(url, "dists"), verify=verify)
+    except HTTPError:
         raise
 
     links = re.findall(r"href=[\"'](.*)[\"']", resp.text)
     return {dist.strip("/") for dist in links if ".." not in dist}
 
 
-def _packages_file(base_url: str) -> str:
+def _packages_file(base_url: str, verify: Optional[str] = None) -> str:
+    """Retrieve the packages file. It is possible for it to be compressed."""
     try:
-        resp = get_url(urljoin(base_url, "Packages"))
+        resp = get_url(urljoin(base_url, "Packages"), verify=verify)
         return resp.text
     except HTTPError as e:
         if e.response.status_code == 404:
-            resp = get_url(urljoin(base_url, "Packages.gz"))
+            resp = get_url(urljoin(base_url, "Packages.gz"), verify=verify)
             return zlib.decompress(resp.content, 16 + zlib.MAX_WBITS).decode()
         else:
             raise e
 
 
-def check_apt_repo(url: str, dists: Optional[Set[str]]) -> bool:
+def _check_apt_repo_metadata(url: str, dist: str, release_file: Release,
+                             errors: RepoErrors, verify: Optional[str] = None) -> None:
+    """Check repo metadata for checksum mismatches."""
+
+    dist_url = urljoin(url, "dists", dist)
+    checksum_types = {key: CHECKSUMS[key] for key in CHECKSUMS.keys() if key in release_file}
+
+    success = True
+
+    files = dict()
+    for key in checksum_types:
+        for file_def in release_file[key]:
+            filename = file_def['name']
+            if filename not in files:
+                files[filename] = dict()
+            files[filename][key] = file_def[key.lower()]
+
+    if not files:
+        errors.add(
+            url, dist,
+            urljoin(dist_url, "Release") + " file malformed"
+        )
+        success = False
+
+    for file_name, checksums in files.items():
+        multihash = MultiHash(list(checksum_types.values()))
+
+        file_url = urljoin(dist_url, file_name)
+        try:
+            response = get_url(f"{file_url}", verify=verify, stream=True)
+        except HTTPError as e:
+            if f"{file_name}.gz" not in files:
+                errors.add(
+                    url, dist,
+                    f"Could not access file at {e.respose.url} : {e}"
```

Comment:
I think this should maybe be `response`?

Reply:
Yeah good catch